Typescript migration

berekuk opened the pull request:
This is going to be a long branch (~2 weeks, maybe? I hope to finish it by the start of my short vacation, i.e. by December 8th).

I'm taking a very conservative bottom-up approach: I'm rewriting every file function-by-function with the minimal amount of improvements and intermediate Rescript->Typescript bindings. It'll be easier to improve on the codebase after the conversion is done.

There are some downsides to it:
- it's BORING
- sometimes I have to introduce hacks, e.g. drop generics (it's hard to combine Rescript generics and `js.raw` calls) or build `result` objects by hand with their `TAG: 0` and `_0: ...` fields
- it's possible to make a typo when I declare Rescript->JS bindings

But these problems (except for boredom) are temporary and will fix themselves after I'm done with this branch. And it's more predictable; my initial attempt to rewrite distributions with classes failed due to the total number of changes I introduced simultaneously.

- [X] convert `XYShape.res` (~1 day of work, done)
- [x] convert `PointSetDist/*.res` (2-3 days?)
- [x] rewrite XYShape and Continuous/Discrete/Mixed in a more Typescript-idiomatic way
- [x] convert GenericDist and hide it behind an opaque type (`IEvDistribution(dist)`, where `dist` is an instance of Typescript class)
- [x] interpreter / FunctionRegistry rewrite - will figure out the details for this step later
- [x] update docs
- [x] fix rsResult mess
- [x] move files out of `js/`
- [x] naming conventions

OAGr commented:
BTW, the strategy I was thinking of, if I were to do it, would be to begin with a separate JS library for the pointset stuff. Test it more than the Rescript lib (which isn't saying much). This would take more time though, so I'm happy to see either way done.

berekuk commented:
Ok, I converted almost all of PointSetDist code (and also XYShape, and Sparklines, and parts of Operations, and some bits of E). Took me less time than I expected!

`pointSetDist` and all its subtypes are opaque in Rescript, all operations are proxied to the Typescript code.

Typescript code right now is a bit ugly, but I hope to improve on it before merging this. (My ultimate bonus goal here is to have the negative diff; Typescript is more verbose, but we had a lot of duplication in PointSetDist that could be reduced with classes and inheritance and smarter codebase organization; also, I'll be able to drop ForTS after the rewrite is complete).

Performance is 5% better already, btw. Don't know why, maybe because of Rescript's currying overhead.

I'm sure there are bugs, but all tests are passing.
- The main source of bugs is Rescript<->Typescript boundary, which I cross with unsafe `%raw(...)` statements. But this is temporary and will be type-checked when I convert the rest of code.
- Another is typos during the rewrite; tests and type checks catch most of these, but some probably got through. There's no obvious way to catch all of these, except for covering the entire codebase with tests, which I don't want to do yet. I'll write more tests when I get to the refactoring phase.

berekuk commented:
SampleSetDist rewrite is in.

Performance improvements are now more significant, 80s -> 56s.

That's probably not because of Typescript but because of other changes, e.g. `samplesMap`, `map2`, `map3` and `mapN` don't produce arrays of results anymore. This is the my main suspect for the speedup, since these functions were also used in `minOfTwo` and `minOfFloat` and so on.

berekuk commented:
My initial conversion used naive immutable implementation (wrapper class that copied the native `Map` on every change).

In the last commit switched the implementation to Immutable.js. Not surprisingly, Immutable.js is slower on tiny benchmarks because of constant overhead.

Surprisingly, overall result is slower both on full GUCEM and on the synthetic `bench-map.ts`.

I.e., full copy of bindings on each change is faster than going through Immutable.js.

`bench-map.ts` is even slower than it was on 0.5 (`bench-map-reduce` is faster, so this is noticeable only on very synthetic benchmarks). But I'm not sure if Immutable.js is to blame here or some other changes, e.g. `Value` shapes which I haven't optimized yet (value types are currently strings, not numbers as they were in Rescript, but I intend to replace them with enums).

Overall, though, I think we still should go with Immutable.js, since it would scale better. `Belt.Map` implementation would be worth considering too, we could copy it over from Rescript. I couldn't find any other popular implementations of immutable maps in JS, unfortunately (well, there's immer.js, but it's too magical for my taste, and it's [slower](https://immerjs.github.io/immer/performance/) than immutable.js).

berekuk commented:
Highlights of this PR that I haven't mentioned yet:

1. FunctionRegistry uses unpacked values! `frTypes` now consists of generic functions that are responsible both for checking whether the value matches the signature and for returning the unpacked values. ImpossibleErrors are no more.
2. There's a single `FnFactory` class for building functions, with many `n2n`, `nn2n`, `d2d` etc methods that help to build functions with even less code. `makeDefinition` is a more general fallback when a required signature doesn't have a helper method. (I dislike our `FRFunction` vs `fnDefinition` duality, btw, but I didn't refactor it away yet.)
3. Distributions are now classes, based on the abstract `BaseDist` base class. We don't have `GenericDist` ADT anymore, and dist subtypes are checked with `instanceof`.
4. `result` type is `{ ok: boolean, value: ... }`, not `{ tag: "Yes" | "No", value: ... }`. This is a breaking change that I had to do in `components` and `vscode-ext` too, but I believe this shape is slightly better / more natural.

Hazelfire commented:
Just did some testing, and I got a 4x speed increase on my observable work. So I got something on my end. Thanks for the good work!

Not sure how I am to go about reviewing such a massive piece of work, I tried it in observable, and it seems to work fine. I'm happy to see this merged.

berekuk commented:
I'm going to merge this.

Changes I added in the last few days:
- converted some types to classes (unfinished; XYShape, Result, ErrorMessage, Expression, AST and others are still plain objects)
- did some minor performance optimizations, e.g. added array preallocations and replaced map loops with for loops when it mattered in benchmarks (I tried both to optimize against some long squiggle scripts and get to the point where scripts/bench-map.ts is not slower than 0.5)
- renamed files to lowerCase.ts and reorganized some tests

There's still a lot to be done, but this is stable enough to continue on `develop` branch and to release a preview npm build.
